### ISSUE TITLE
Exposing PyVista's filters as vtkAlgorithms

### ISSUE THREAD

**Issue by banesullivan** [pyvista/pyvista#3809]
#3318 is soon to introduce broad `vtkAlgorithm` support to `BasePlotter.add_mesh()` (and a few other methods). While initially, this is intended to be used by more advanced users, I'd like to move towards a pipeline filtering API that is approachable to the average user.

I'm trying to brainstorm some ways to go about this and would like to solicit community involvement.

## `return_algorithm` keyword argument 

We could introduce a `return_algorithm` keyword argument for all filter methods that will return the `vtkAlgorithm` for that filter. This would allow a user to configure the algorithm to process their data the way they like and connect to a pipeline. In practice, this may look like:

```py
alg = mesh.slice(return_algorithm=True)

def update_mesh(*args):
    mesh['scalars'] = np.random.random(mesh.n_points)

pl = pv.Plotter()
pl.add_mesh(alg)
pl.add_key_event('t', update_mesh)
pl.show()
# User hits `t` key and the rendered slice will update with the new data
```

This gets tricky when we want to chain operations together, e.g., `mesh.threshold().slice()`. I'm not sure of how to go about designing a sane API for chaining algorithms together like this without a full re-wrtie of the the filters.

## Keep in mind

FWIW, VTK recently introduced an `override` mechanism that will make wrapping filters (`vtkAlgorithm`s) much easier. I see a future where all the VTK filter classes are wrapped in a PyVista-like fashion. 

Perhaps the best solution is to go this route... re-wrapping the filters as Pythonic `vtkAlgorithm`s.

For example:

```py
@vtkContourFilter.override
class Contour(vtkContourFilter):
    def __call__(
        self,
        inp,  # Could be mesh or other algorithm
        isosurfaces=10,
        scalars=None,
        compute_normals=False,
        compute_gradients=False,
        compute_scalars=True,
        rng=None,
        preference='point',
        method='contour',
        progress_bar=False,
    ):
        ...  # Set up algorithm

    def __init__(self, *args, **kwargs):
        super().__init__()
        self(*args, **kwargs)

    ...
    @property.setter
    def isosurfaces(self, n):
        # Set number of isosurfaces
        ...

mesh = pv.Wavelet()
alg = Contour(mesh, scalars=mesh.active_scalars_name)

def set_n(n):
    alg.isosurfaces = n

pl = pv.Plotter()
pl.add_mesh(alg)
pl.add_slider_widget(set_n, (2, 20))
pl.show()
```

**Comment by darikg** [comment]
Does the python subclassing vtk filters like that support mixins or monkey-patching methods into the vtkFilter superclass, so common convenience methods could be added to all filter classes?

> This gets tricky when we want to chain operations together, e.g., mesh.threshold().slice()

not even sure this is a good idea but just throwing it out there, use a context-manager to control return type:

```python
# equivalent to mesh.threshold(return_algorithm=True).slice(return_algorithm=True)
with pv.chain_filters():
  alg = mesh.threshold().slice()
```

**Comment by banesullivan** [comment]
> Does the python subclassing vtk filters like that support mixins or monkey-patching methods into the vtkAlgorithm superclass, so common convenience methods could be added to all filter subclasses?

Yes, in the same way that our subclassing of datasets in PyVista works. and like what's in PyVista right now, those overrides are only available on the Python-side.

Subclassing `vtkAlgorithm` and adding methods will not add those methods to all VTK filters though (mush the same way that subclassing `vtkDataSet` doesn't automagically give all datasets convienance methods). We have to manually override every class want to use that way. For example:

```py
import vtk

@vtk.vtkAlgorithm.override
class AlgorithmBase(vtk.vtkAlgorithm):
    def __init__(self) -> None:
        super().__init__()

    def foo(self):
        return 'foo'

@vtk.vtkContourFilter.override
class Contour(vtk.vtkContourFilter, AlgorithmBase):
    def __init__(self) -> None:
        super().__init__()
```

```py
>>> alg = Contour()
>>> alg.foo()
'foo'
```

**Comment by banesullivan** [comment]
And incase it isn't clear, the `override` decorator makes this possible:

```py
>>> vtk.vtkContourFilter()
<Contour(0x134f7f570) at 0x10e870c40>
```

**Comment by akaszynski** [comment]
First, a nitpick: suggest going with `return_alg` because it's shorter.

> This gets tricky when we want to chain operations together, e.g., mesh.threshold().slice(). I'm not sure of how to go about designing a sane API for chaining algorithms together like this without a full re-wrtie of the the filters.

I think this is where VTK pipelines shines; chaining operations. However, I agree that we would need a full rewrite of the filters. Right now, our filters assume they're operating on `pyvista.DataSet` objects and often use explicit information from the input dataset. We would be forced to execute the algorithm and pass the data onward.

I think one way to work around this would be to wrap the filter classes themselves and not treat them as functions (as implemented within the `DataSet` classes). Chaining operations would then work "out of the box" in a "pythonic" manner.

**Comment by MatthewFlamm** [comment]
One thing this could make possible is the use of time dependent pipelines.  One use case is using a (set of) file(s) with time dependent data, then you could envision having a slider widget for time.  There are also filters that only work in pipeline mode with time dependent sources.

Id like this to be part of the design thinking so that we don't have to overhaul the structure a second time to enable it.

The complexity of this in vtk is what originally drew me to PyVista.  I didn't have to become an expert in vtk pipelines just to iterate over time and run a quick analysis.  I agree with other suggestions that it would be great to have flexibility to do the things I wrote above, but somehow with the "just-working" (my words) methodology of PyVista.

**Comment by banesullivan** [comment]
@darikg, I really like that approach! Thanks for putting that together!
